test-support/reference-trie/src/lib.rs

```diff
@@ -17,21 +17,153 @@
 use std::fmt;
 use std::error::Error as StdError;
 use std::iter::once;
+use std::marker::PhantomData;
 use parity_codec::{Decode, Input, Output, Encode, Compact};
 use trie_root::Hasher;
-use trie_db::{node::Node, triedbmut::ChildReference, DBValue};
+use trie_db::{
+	node::Node,
+	triedbmut::ChildReference,
+	DBValue,
+	trie_visit,
+	TrieBuilder,
+	TrieRoot,
+	Partial,
+	Cache16,
+	Cache4,
+};
+use std::borrow::Borrow;
 use keccak_hasher::KeccakHasher;
 
-pub use trie_db::{Trie, TrieMut, NibbleSlice, NodeCodec, Recorder, Record};
+pub use trie_db::{Trie, TrieMut, NibbleSlice, Recorder, NodeCodec, ChildBitmap,
+	ChildSliceIx};
+pub use trie_db::{Record, TrieLayOut, TrieOps, NibbleHalf, NibbleQuarter, NibbleOps};
 pub use trie_root::TrieStream;
 
-pub type RefTrieDB<'a> = trie_db::TrieDB<'a, keccak_hasher::KeccakHasher, ReferenceNodeCodec>;
-pub type RefTrieDBMut<'a> = trie_db::TrieDBMut<'a, KeccakHasher, ReferenceNodeCodec>;
-pub type RefFatDB<'a> = trie_db::FatDB<'a, KeccakHasher, ReferenceNodeCodec>;
-pub type RefFatDBMut<'a> = trie_db::FatDBMut<'a, KeccakHasher, ReferenceNodeCodec>;
-pub type RefSecTrieDB<'a> = trie_db::SecTrieDB<'a, KeccakHasher, ReferenceNodeCodec>;
-pub type RefSecTrieDBMut<'a> = trie_db::SecTrieDBMut<'a, KeccakHasher, ReferenceNodeCodec>;
-pub type RefLookup<'a, Q> = trie_db::Lookup<'a, KeccakHasher, ReferenceNodeCodec, Q>;
+/// trie layout similar to parity-ethereum
+pub struct LayoutOri;
```

Comment:
`LayoutOrigin`? Full name would be more readable

Reply:
Yes, I am also removing the idea of new and origin, I will use 'ExtensionLayout' and 'NoExtensionLayout'. The fact that reference trie uses same scheme as substrate should not be something strict (code will diverge at some point).